lib/Jsrt/JsrtDebugUtils.cpp

```diff
@@ -0,0 +1,465 @@
+        utf8::DecodeIntoAndNullTerminate(sourceContent, source + startByte, byteLength, options);
+        JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::sourceText, sourceContent, byteLength, functionBody->GetScriptContext());
+    }
+    else
+    {
+        JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::sourceText, _u(""), 1, functionBody->GetScriptContext());
+    }
+}
+
+void JsrtDebugUtils::AddLineCountToObject(Js::DynamicObject * object, Js::Utf8SourceInfo * utf8SourceInfo)
+{
+    utf8SourceInfo->EnsureLineOffsetCache();
+
+    JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::lineCount, (ULONG)utf8SourceInfo->GetLineCount(), utf8SourceInfo->GetScriptContext());
+}
+
+void JsrtDebugUtils::AddSouceToObject(Js::DynamicObject * object, Js::Utf8SourceInfo * utf8SourceInfo)
+{
+    int32 cchLength = utf8SourceInfo->GetCchLength();
+    AutoArrayPtr<char16> sourceContent(HeapNewNoThrowArray(char16, cchLength + 1), cchLength + 1);
+    if (sourceContent != nullptr)
+    {
+        utf8::DecodeOptions options = utf8SourceInfo->IsCesu8() ? utf8::doAllowThreeByteSurrogates : utf8::doDefault;
+        utf8::DecodeIntoAndNullTerminate(sourceContent, utf8SourceInfo->GetSource(), cchLength, options);
+        JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::source, sourceContent, cchLength, utf8SourceInfo->GetScriptContext());
+    }
+    else
+    {
+        JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::source, _u(""), 1, utf8SourceInfo->GetScriptContext());
+    }
+}
+
+void JsrtDebugUtils::AddVarPropertyToObject(Js::DynamicObject * object, const char16 * propertyName, Js::Var value, Js::ScriptContext * scriptContext)
+{
+    const Js::PropertyRecord* propertyRecord;
+
+    // propertyName is the DEBUGOBJECTPROPERTY from JsrtDebugPropertiesEnum so it can't have embedded null, ok to use wcslen
+    scriptContext->GetOrAddPropertyRecord(propertyName, static_cast<int>(wcslen(propertyName)), &propertyRecord);
+
+    Js::Var marshaledObj = Js::CrossSite::MarshalVar(scriptContext, value);
+
+    if (FALSE == Js::JavascriptOperators::InitProperty(object, propertyRecord->GetPropertyId(), marshaledObj))
+    {
+        Assert("Failed to add property to debugger object");
+    }
+}
+
+void JsrtDebugUtils::AddPropertyType(Js::DynamicObject * object, Js::IDiagObjectModelDisplay* objectDisplayRef, Js::ScriptContext * scriptContext)
+{
+    Assert(objectDisplayRef != nullptr);
+    Assert(scriptContext != nullptr);
+
+    bool addDisplay = false;
+    bool addValue = false;
+
+    Js::Var varValue = objectDisplayRef->GetVarValue(FALSE);
+
+    if (varValue != nullptr)
+    {
+        Js::TypeId typeId = Js::JavascriptOperators::GetTypeId(varValue);
+
+        switch (typeId)
+        {
+        case Js::TypeIds_Undefined:
+            JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::type, scriptContext->GetLibrary()->GetUndefinedDisplayString(), scriptContext);
```

Comment:
I am fine - but our both debuggers are differing in output. What does Limin say about it?

---
In reply to: [62752235](https://github.com/Microsoft/ChakraCore/pull/926#discussion_r62752235) [](ancestors = 62752235)